File: packages/sdk-ts/src/tools.ts
// suite.tools.* — Model Context Protocol (MCP) server + tool operations.
//
// Endpoint paths and JSON shapes are the canonical contract from
// `apps/dashboard/src/lib/api.ts` (the Phase 11.1 MCP section). Schemas
// below mirror the zod schemas there; if the dashboard contract moves,
// move these together.
//
// Public API stays camelCase; the wire stays snake_case. We translate at
// the boundary — `command`, `env`, `arguments`, and `content` are
// opaque caller-defined payloads that pass through verbatim.

import { z } from "zod"
import { request, type HttpOptions } from "./_http.js"

// ---------- shared schemas (mirror lib/api.ts, camelCase) ----------

export const MCPTransportSchema = z.enum(["stdio", "sse"])
export type MCPTransport = z.infer<typeof MCPTransportSchema>

export const MCPServerStatusSchema = z.enum([
  "connecting",
  "ready",
  "errored",
  "disabled",
])
export type MCPServerStatus = z.infer<typeof MCPServerStatusSchema>

export const MCPServerSchema = z.object({
  name: z.string(),
  transport: MCPTransportSchema,
  command: z.array(z.string()),
  url: z.string().nullable(),
  env: z.record(z.string(), z.string()),
  tenantId: z.string().nullable(),
  description: z.string(),
  isEnabled: z.boolean(),
  status: MCPServerStatusSchema,
  lastError: z.string().nullable(),
  toolsCount: z.number(),
  installedAt: z.string(),
  lastConnectedAt: z.string().nullable(),
})
export type MCPServer = z.infer<typeof MCPServerSchema>

export const MCPServerListSchema = z.object({
  servers: z.array(MCPServerSchema),
})
export type MCPServerList = z.infer<typeof MCPServerListSchema>

export const MCPToolSchema = z.object({
  id: z.string(),
  server: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  inputSchema: z.record(z.string(), z.unknown()),
})
export type MCPTool = z.infer<typeof MCPToolSchema>

export const MCPToolListSchema = z.object({
  tools: z.array(MCPToolSchema),
})
export type MCPToolList = z.infer<typeof MCPToolListSchema>

export const MCPCallResultSchema = z.object({
  content: z.array(z.record(z.string(), z.unknown())),
  isError: z.boolean(),
  durationMs: z.number(),
})
export type MCPCallResult = z.infer<typeof MCPCallResultSchema>

// ---------- option shapes ----------

export interface AddMCPServerOptions extends HttpOptions {
  /** stdio: ["uvx", "mcp-server-github"]. Required when transport=stdio. */
  command?: string[]
  /** sse: "https://mcp.acme.com/sse". Required when transport=sse. */
  url?: string
  /**
   * Environment passed to stdio servers. Values can reference vault keys
   * via the `"secret:<key>"` prefix — never raw secret strings.
   */
  env?: Record<string, string>
  description?: string
  /** Per-tenant scoping. Omit = available to every tenant. */
  tenantId?: string
}

export interface ListMCPToolsOptions extends HttpOptions {
  /** Scope the listing to one server. */
  server?: string
}

const ALLOWED_TRANSPORTS: readonly MCPTransport[] = ["stdio", "sse"]

// ---------- public API ----------

/** List every configured MCP server visible to the caller. */
export async function listMcpServers(
  opts: HttpOptions = {},
): Promise<MCPServer[]> {
  const raw = await request<unknown>("GET", "/mcp/servers", null, opts)
  const parsed = MCPServerListSchema.parse(camelizeServerList(raw))
  return parsed.servers
}

/**
 * Register a new MCP server.
 *
 * For `transport: "stdio"` you must supply `command`; for
 * `transport: "sse"` you must supply `url`. The runtime will attempt to
 * connect immediately and the returned `status` will reflect the first
 * probe outcome (`ready` / `errored`).
 */
export async function addMcpServer(
  name: string,
  transport: MCPTransport,
  opts: AddMCPServerOptions = {},
): Promise<MCPServer> {
  if (typeof name !== "string" || name.length === 0) {
    throw new Error("mcp server name must be a non-empty string")
  }
  if (!(ALLOWED_TRANSPORTS as readonly string[]).includes(transport)) {
    throw new Error(
      `transport must be one of: ${ALLOWED_TRANSPORTS.join(", ")}; got ${JSON.stringify(transport)}`,
    )
  }
  const { command, url, env, description, tenantId, ...http } = opts

  if (transport === "stdio") {
    if (!Array.isArray(command) || command.length === 0) {
      throw new Error("stdio transport requires a non-empty command array")
    }
    for (const arg of command) {
      if (typeof arg !== "string") {
        throw new Error("command entries must all be strings")
      }
    }
  }
  if (transport === "sse") {
    if (typeof url !== "string" || url.length === 0) {
      throw new Error("sse transport requires a url")
    }
  }

  const body: Record<string, unknown> = { name, transport }
  if (command !== undefined) body.command = [...command]
  if (url !== undefined) body.url = url
  if (env !== undefined) body.env = { ...env }
  if (description !== undefined) body.description = description
  if (tenantId !== undefined) body.tenant_id = tenantId

  const raw = await request<unknown>("POST", "/mcp/servers", body, http)
  return MCPServerSchema.parse(camelizeServer(raw))
}

/**
 * Remove an MCP server. The runtime closes the underlying connection
 * and the row is deleted.
 */
export async function removeMcpServer(
  name: string,
  opts: HttpOptions = {},
): Promise<void> {
  if (typeof name !== "string" || name.length === 0) {
    throw new Error("mcp server name must be a non-empty string")
  }
  await request<unknown>(
    "DELETE",
    `/mcp/servers/${encodeURIComponent(name)}`,
    null,
    opts,
  )
}

/**
 * Enable or disable an MCP server. Disabling closes the connection but
 * preserves the configuration.
 */
export async function enableMcpServer(
  name: string,
  enabled: boolean,
  opts: HttpOptions = {},
): Promise<MCPServer> {
  if (typeof name !== "string" || name.length === 0) {
    throw new Error("mcp server name must be a non-empty string")
  }
  const raw = await request<unknown>(
    "PUT",
    `/mcp/servers/${encodeURIComponent(name)}/enabled`,
    { enabled: Boolean(enabled) },
    opts,
  )
  return MCPServerSchema.parse(camelizeServer(raw))
}

/** List MCP tools, optionally scoped to a single server. */
export async function listMcpTools(
  opts: ListMCPToolsOptions = {},
): Promise<MCPTool[]> {
  const { server, ...http } = opts
  if (server !== undefined && server.length === 0) {
    throw new Error("server filter must be a non-empty string")
  }
  const qs = new URLSearchParams()
  if (server !== undefined) qs.set("server", server)
  const qsStr = qs.toString()
  const path = qsStr.length > 0 ? `/mcp/tools?${qsStr}` : "/mcp/tools"
  const raw = await request<unknown>("GET", path, null, http)
  const parsed = MCPToolListSchema.parse(camelizeToolList(raw))
  return parsed.tools
}

/**
 * Invoke an MCP tool and return its result. The returned `content`
 * array carries the MCP protocol's typed parts (text / image / json) —
 * caller code should check `isError` before consuming.
 */
export async function callMcp(
  server: string,
  tool: string,
  args?: Record<string, unknown>,
  opts: HttpOptions = {},
): Promise<MCPCallResult> {
  if (typeof server !== "string" || server.length === 0) {
    throw new Error("mcp server name must be a non-empty string")
  }
  if (typeof tool !== "string" || tool.length === 0) {
    throw new Error("mcp tool name must be a non-empty string")
  }
  const body: Record<string, unknown> = { server, tool }
  if (args !== undefined) body.arguments = { ...args }
  const raw = await request<unknown>("POST", "/mcp/call", body, opts)
  return MCPCallResultSchema.parse(camelizeCallResult(raw))
}

// ---------- helpers ----------

function camelizeServer(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  return {
    name: src.name,
    transport: src.transport,
    command: src.command ?? [],
    url: src.url ?? null,
    env: src.env ?? {},
    tenantId: src.tenant_id ?? src.tenantId ?? null,
    description: src.description ?? "",
    isEnabled: src.is_enabled ?? src.isEnabled,
    status: src.status,
    lastError: src.last_error ?? src.lastError ?? null,
    toolsCount: src.tools_count ?? src.toolsCount,
    installedAt: src.installed_at ?? src.installedAt,
    lastConnectedAt: src.last_connected_at ?? src.lastConnectedAt ?? null,
  }
}

function camelizeServerList(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  const servers = Array.isArray(src.servers)
    ? src.servers.map(camelizeServer)
    : []
  return { servers }
}

function camelizeTool(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  return {
    id: src.id,
    server: src.server,
    name: src.name,
    description: src.description ?? null,
    // `input_schema` is forwarded as an opaque JSON Schema object — we
    // only rename the wrapping key, not any inner field names.
    inputSchema: src.input_schema ?? src.inputSchema ?? {},
  }
}

function camelizeToolList(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  const tools = Array.isArray(src.tools) ? src.tools.map(camelizeTool) : []
  return { tools }
}

function camelizeCallResult(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  return {
    // `content` parts are opaque MCP payloads — pass through unchanged.
    content: Array.isArray(src.content) ? src.content : [],
    isError: src.is_error ?? src.isError ?? false,
    durationMs: src.duration_ms ?? src.durationMs ?? 0,
  }
}

/** Namespace object — the shape `suite.tools` is built from. */
export const tools = {
  listMcpServers,
  addMcpServer,
  removeMcpServer,
  enableMcpServer,
  listMcpTools,
  callMcp,
} as const

File: packages/sdk-py/af_stack/tools.py
# SPDX-License-Identifier: Apache-2.0

"""``suite.tools.*`` — Model Context Protocol (MCP) server + tool operations.

Maps to the REST surface defined in ``apps/dashboard/src/lib/api.ts`` under
the Phase 11.1 MCP section:

==================================  ========================================
SDK call                            Endpoint
==================================  ========================================
:func:`list_mcp_servers`            ``GET    /api/v1/mcp/servers``
:func:`add_mcp_server`              ``POST   /api/v1/mcp/servers``
:func:`remove_mcp_server`           ``DELETE /api/v1/mcp/servers/{name}``
:func:`enable_mcp_server`           ``PUT    /api/v1/mcp/servers/{name}/enabled``
:func:`list_mcp_tools`              ``GET    /api/v1/mcp/tools[?server=...]``
:func:`call_mcp`                    ``POST   /api/v1/mcp/call``
==================================  ========================================

MCP servers are external processes (stdio) or HTTP/SSE endpoints that
expose tools to agents. The suite runtime hosts a pool of connections
— one per configured server — and proxies tool calls. Per-tenant
isolation: when MT is on, each tenant can opt servers in/out via the
dashboard.

The :class:`MCPServer`, :class:`MCPTool`, and :class:`MCPCallResult`
Pydantic models mirror the zod schemas in ``lib/api.ts`` exactly — keep
them in sync when the contract changes.

.. note::
   This module was originally exposed as a ``Tools`` class with bound
   methods. The class is kept as a thin backward-compat shim and will be
   removed in a future major release — callers should use the
   module-level functions (or the ``suite.tools.<verb>`` access form).
"""

from __future__ import annotations

import warnings
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from . import _http

MCPTransport = Literal["stdio", "sse"]
MCPServerStatus = Literal["connecting", "ready", "errored", "disabled"]
ToolAdapterID = Literal["browser-use", "searxng", "fs", "exec", "http", "sql"]
NativeToolName = Literal["browser", "search", "fs", "exec", "http", "sql"]

_ALLOWED_TRANSPORTS: tuple[str, ...] = ("stdio", "sse")


class MCPServer(BaseModel):
    """One configured MCP server row.

    Mirrors ``MCPServerSchema`` in ``apps/dashboard/src/lib/api.ts``.
    """

    model_config = ConfigDict(extra="allow")

    name: str
    transport: MCPTransport
    command: list[str] = Field(default_factory=list)
    url: str | None = None
    env: dict[str, str] = Field(default_factory=dict)
    tenant_id: str | None = None
    description: str = ""
    is_enabled: bool = True
    status: MCPServerStatus = "connecting"
    last_error: str | None = None
    tools_count: int = 0
    installed_at: str
    last_connected_at: str | None = None


class MCPTool(BaseModel):
    """One tool exposed by an MCP server.

    Mirrors ``MCPToolSchema`` in ``apps/dashboard/src/lib/api.ts``.
    """

    model_config = ConfigDict(extra="allow")

    id: str
    server: str
    name: str
    description: str | None = None
    input_schema: dict[str, Any] = Field(default_factory=dict)


class MCPCallResult(BaseModel):
    """Result of an MCP tool call.

    Mirrors ``CallMCPResultSchema`` in ``apps/dashboard/src/lib/api.ts``.
    The ``content`` array carries the MCP protocol's typed content parts
    (text / image / json / etc.) — kept opaque so callers can pull
    whichever parts they need.
    """

    model_config = ConfigDict(extra="allow")

    content: list[dict[str, Any]] = Field(default_factory=list)
    is_error: bool = False
    duration_ms: int = 0


class BuiltinToolDefinition(BaseModel):
    """One tool exposed by a built-in AF Stack adapter."""

    model_config = ConfigDict(extra="allow")

    name: str
    description: str = ""
    input_schema: dict[str, Any] = Field(default_factory=dict)


class ToolAdapter(BaseModel):
    """Built-in adapter catalogue row plus tenant enablement."""

    model_config = ConfigDict(extra="allow")

    id: ToolAdapterID
    label: str
    description: str = ""
    enabled: bool = False
    configured: bool = False
    default_enabled: bool = False
    config: dict[str, Any] = Field(default_factory=dict)
    updated_at: str | None = None
    tools: list[BuiltinToolDefinition] = Field(default_factory=list)


class ToolAdapterCallResult(BaseModel):
    """Result from ``call_adapter``."""

    model_config = ConfigDict(extra="allow")

    adapter: ToolAdapterID
    tool: str
    status: str
    result: Any | None = None
    duration_ms: int = 0


# ─── Native tools (#16 — strict-interface adapter set) ───────────────────


class NativeToolVerb(BaseModel):
    """One verb exposed by a native tool (input/output schemas)."""

    model_config = ConfigDict(extra="allow")

    name: str
    description: str = ""
    input_schema: dict[str, Any] = Field(default_factory=dict)
    output_schema: dict[str, Any] | None = None


class NativeToolStatus(BaseModel):
    """Per-tenant view of one native tool: active adapter + enable flag."""

    model_config = ConfigDict(extra="allow")

    tool: NativeToolName
    description: str = ""
    adapter_id: str
    configured: bool = False
    enabled: bool = False
    verbs: list[NativeToolVerb] = Field(default_factory=list)
    config: dict[str, Any] = Field(default_factory=dict)
    updated_at: str | None = None


class InvokeNativeToolResult(BaseModel):
    """Result from ``invoke_native``."""

    model_config = ConfigDict(extra="allow")

    tool: NativeToolName
    verb: str
    result: Any | None = None


# ─── Module-level verbs ───────────────────────────────────────────────────


async def list_mcp_servers() -> list[MCPServer]:
    """Return every configured MCP server visible to the caller.

    When multi-tenancy is on, the runtime scopes the list to the active
    tenant; servers with ``tenant_id is None`` (shared catalogue) appear
    for every tenant.
    """
    body = await _http.request_json("GET", "/mcp/servers")
    payload = body or {"servers": []}
    servers_raw = payload.get("servers") if isinstance(payload, dict) else None
    if not isinstance(servers_raw, list):
        return []
    return [MCPServer.model_validate(row) for row in servers_raw]


async def add_mcp_server(
    name: str,
    transport: MCPTransport | str,
    *,
    command: list[str] | None = None,
    url: str | None = None,
    env: dict[str, str] | None = None,
    description: str | None = None,
    tenant_id: str | None = None,
) -> MCPServer:
    """Register a new MCP server.

    For ``transport="stdio"`` you must supply ``command``; for
    ``transport="sse"`` you must supply ``url``. The runtime will attempt
    to connect immediately and the returned ``status`` will reflect the
    first probe outcome (``ready`` / ``errored``).

    ``env`` entries may reference secrets vault keys via the
    ``"secret:<key>"`` prefix — the runtime resolves them at process
    spawn time so raw secret values never travel over the wire.
    """
    if not name:
        raise ValueError("mcp server name must be a non-empty string")
    if transport not in _ALLOWED_TRANSPORTS:
        allowed = ", ".join(_ALLOWED_TRANSPORTS)
        raise ValueError(
            f"transport must be one of: {allowed}; got {transport!r}"
        )
    if transport == "stdio":
        if not command:
            raise ValueError("stdio transport requires a non-empty command list")
        for arg in command:
            if not isinstance(arg, str):
                raise ValueError("command entries must all be strings")
    if transport == "sse":
        if not url:
            raise ValueError("sse transport requires a url")

    payload: dict[str, Any] = {"name": name, "transport": transport}
    if command is not None:
        payload["command"] = list(command)
    if url is not None:
        payload["url"] = url
    if env is not None:
        payload["env"] = dict(env)
    if description is not None:
        payload["description"] = description
    if tenant_id is not None:
        payload["tenant_id"] = tenant_id

    body = await _http.request_json("POST", "/mcp/servers", json=payload)
    return MCPServer.model_validate(body or {})


async def remove_mcp_server(name: str) -> None:
    """Remove an MCP server.

    The runtime closes the underlying connection and the row is deleted.
    In-flight tool calls against the server raise an ``MCP_SERVER_GONE``
    error envelope.
    """
    if not name:
        raise ValueError("mcp server name must be a non-empty string")
    await _http.request_json("DELETE", f"/mcp/servers/{name}")


async def enable_mcp_server(name: str, enabled: bool) -> MCPServer:
    """Enable or disable an MCP server.

    Disabling closes the connection but preserves the configuration —
    the server can be re-enabled later without re-registering it.
    """
    if not name:
        raise ValueError("mcp server name must be a non-empty string")
    body = await _http.request_json(
        "PUT",
        f"/mcp/servers/{name}/enabled",
        json={"enabled": bool(enabled)},
    )
    return MCPServer.model_validate(body or {})


async def list_mcp_tools(server: str | None = None) -> list[MCPTool]:
    """List MCP tools, optionally scoped to a single server.

    The returned ``input_schema`` field is the raw JSON Schema the MCP
    server declared — forwarded unchanged so agents can use it for
    tool-use prompting.
    """
    params: dict[str, Any] = {}
    if server is not None:
        if not server:
            raise ValueError("server filter must be a non-empty string")
        params["server"] = server
    body = await _http.request_json(
        "GET",
        "/mcp/tools",
        params=params if params else None,
    )
    payload = body or {"tools": []}
    tools_raw = payload.get("tools") if isinstance(payload, dict) else None
    if not isinstance(tools_raw, list):
        return []
    return [MCPTool.model_validate(row) for row in tools_raw]


async def call_mcp(
    server: str,
    tool: str,
    arguments: dict[str, Any] | None = None,
) -> MCPCallResult:
    """Invoke an MCP tool and return its result.

    ``arguments`` must conform to the tool's ``input_schema``; the
    runtime forwards the dict verbatim to the MCP server. The returned
    :class:`MCPCallResult` carries the protocol's typed ``content``
    parts (text / image / json) and an ``is_error`` flag — caller code
    should check the flag before consuming content.
    """
    if not server:
        raise ValueError("mcp server name must be a non-empty string")
    if not tool:
        raise ValueError("mcp tool name must be a non-empty string")
    payload: dict[str, Any] = {"server": server, "tool": tool}
    if arguments is not None:
        payload["arguments"] = dict(arguments)
    body = await _http.request_json("POST", "/mcp/call", json=payload)
    return MCPCallResult.model_validate(body or {})


async def list_adapters() -> list[ToolAdapter]:
    """Return built-in adapters and tenant enablement/config.

    This is distinct from MCP servers: these adapters are the AF Stack
    built-ins (HTTP, SQL, exec, fs, SearXNG, browser-use). AgentField still
    owns agent tool-call spans and traces.
    """
    body = await _http.request_json("GET", "/tools/adapters")
    payload = body or {"adapters": []}
    rows = payload.get("adapters") if isinstance(payload, dict) else None
    if not isinstance(rows, list):
        return []
    return [ToolAdapter.model_validate(row) for row in rows]


async def set_adapter_enabled(
    adapter: ToolAdapterID | str,
    enabled: bool,
    *,
    config: dict[str, Any] | None = None,
) -> ToolAdapter:
    """Enable or disable one built-in adapter for the caller tenant."""
    if not adapter:
        raise ValueError("adapter must be a non-empty string")
    payload: dict[str, Any] = {"enabled": bool(enabled)}
    if config is not None:
        payload["config"] = dict(config)
    body = await _http.request_json(
        "PUT",
        f"/tools/adapters/{adapter}/enabled",
        json=payload,
    )
    return ToolAdapter.model_validate(body or {})


async def call_adapter(
    adapter: ToolAdapterID | str,
    tool: str,
    arguments: dict[str, Any] | None = None,
) -> ToolAdapterCallResult:
    """Call an enabled built-in adapter."""
    if not adapter:
        raise ValueError("adapter must be a non-empty string")
    if not tool:
        raise ValueError("tool must be a non-empty string")
    payload: dict[str, Any] = {"adapter": adapter, "tool": tool}
    if arguments is not None:
        payload["arguments"] = dict(arguments)
    body = await _http.request_json("POST", "/tools/call", json=payload)
    return ToolAdapterCallResult.model_validate(body or {})


# ─── Native tool verbs (#16) ──────────────────────────────────────────────


async def list_native() -> list[NativeToolStatus]:
    """Return the six native tools + per-tenant enablement.

    These are distinct from the legacy ``list_adapters`` surface. Native
    tools have a fixed canonical name set (``browser`` / ``search`` /
    ``fs`` / ``exec`` / ``http`` / ``sql``); the active backend per
    type is picked via env (e.g. ``AF_STACK_TOOL_BROWSER``). The verbs
    are accessible to agents under the MCP ``native:<tool>`` namespace.
    """
    body = await _http.request_json("GET", "/tools/native")
    payload = body or {"tools": []}
    rows = payload.get("tools") if isinstance(payload, dict) else None
    if not isinstance(rows, list):
        return []
    return [NativeToolStatus.model_validate(row) for row in rows]


async def set_native_enabled(
    tool: NativeToolName | str,
    enabled: bool,
    *,
    config: dict[str, Any] | None = None,
) -> NativeToolStatus:
    """Enable or disable one native tool for the caller tenant."""
    if not tool:
        raise ValueError("tool must be a non-empty string")
    payload: dict[str, Any] = {"enabled": bool(enabled)}
    if config is not None:
        payload["config"] = dict(config)
    body = await _http.request_json(
        "POST",
        f"/tools/native/{tool}/enable",
        json=payload,
    )
    return NativeToolStatus.model_validate(body or {})


async def invoke_native(
    tool: NativeToolName | str,
    verb: str,
    args: dict[str, Any] | None = None,
) -> InvokeNativeToolResult:
    """Invoke one verb of a native tool.

    This is the direct-call surface used by tests and non-agent
    callers. Agents call native tools through ``app.mcp.call(
    "native:<tool>", "<verb>", {...})`` instead — see the MCP bridge
    documentation for that path.
    """
    if not tool:
        raise ValueError("tool must be a non-empty string")
    if not verb:
        raise ValueError("verb must be a non-empty string")
    payload: dict[str, Any] = {"verb": verb}
    if args is not None:
        payload["args"] = dict(args)
    body = await _http.request_json(
        "POST",
        f"/tools/native/{tool}/invoke",
        json=payload,
    )
    return InvokeNativeToolResult.model_validate(body or {})


# ─── Backward-compat shim ─────────────────────────────────────────────────


class Tools:
    """Deprecated class-based namespace for MCP operations.

    Kept so code that still does ``from af_stack.tools import Tools`` and
    ``tools = Tools()`` keeps working. All methods forward to the
    module-level functions above. New code should use those directly,
    either as ``from af_stack import tools`` or via
    ``suite.tools.<verb>(...)``.

    .. deprecated:: 0.0.2
       The class form will be removed in a future major release.
    """

    def __init__(self) -> None:
        # Instances are stateless; the warning fires once per construction
        # so it appears next to user code rather than at import time.
        warnings.warn(
            "af_stack.tools.Tools is deprecated; use the module-level "
            "functions (`from af_stack import tools; await tools.call_mcp(...)`).",
            DeprecationWarning,
            stacklevel=2,
        )

    async def list_mcp_servers(self) -> list[MCPServer]:
        return await list_mcp_servers()

    async def add_mcp_server(
        self,
        name: str,
        transport: MCPTransport | str,
        *,
        command: list[str] | None = None,
        url: str | None = None,
        env: dict[str, str] | None = None,
        description: str | None = None,
        tenant_id: str | None = None,
    ) -> MCPServer:
        return await add_mcp_server(
            name,
            transport,
            command=command,
            url=url,
            env=env,
            description=description,
            tenant_id=tenant_id,
        )

    async def remove_mcp_server(self, name: str) -> None:
        return await remove_mcp_server(name)

    async def enable_mcp_server(self, name: str, enabled: bool) -> MCPServer:
        return await enable_mcp_server(name, enabled)

    async def list_mcp_tools(self, server: str | None = None) -> list[MCPTool]:
        return await list_mcp_tools(server)

    async def call_mcp(
        self,
        server: str,
        tool: str,
        arguments: dict[str, Any] | None = None,
    ) -> MCPCallResult:
        return await call_mcp(server, tool, arguments)

    async def list_adapters(self) -> list[ToolAdapter]:
        return await list_adapters()

    async def set_adapter_enabled(
        self,
        adapter: ToolAdapterID | str,
        enabled: bool,
        *,
        config: dict[str, Any] | None = None,
    ) -> ToolAdapter:
        return await set_adapter_enabled(adapter, enabled, config=config)

    async def call_adapter(
        self,
        adapter: ToolAdapterID | str,
        tool: str,
        arguments: dict[str, Any] | None = None,
    ) -> ToolAdapterCallResult:
        return await call_adapter(adapter, tool, arguments)


__all__ = [
    "MCPCallResult",
    "MCPServer",
    "MCPServerStatus",
    "MCPTool",
    "MCPTransport",
    "BuiltinToolDefinition",
    "InvokeNativeToolResult",
    "NativeToolName",
    "NativeToolStatus",
    "NativeToolVerb",
    "Tools",
    "ToolAdapter",
    "ToolAdapterCallResult",
    "ToolAdapterID",
    "add_mcp_server",
    "call_adapter",
    "call_mcp",
    "enable_mcp_server",
    "invoke_native",
    "list_adapters",
    "list_mcp_servers",
    "list_mcp_tools",
    "list_native",
    "remove_mcp_server",
    "set_adapter_enabled",
    "set_native_enabled",
]
